src/SignalR/common/Protocols.Json/src/Protocol/JsonHubProtocol.cs

```diff
@@ -0,0 +1,777 @@
+            WriteInvocationId(message, ref writer);
+            if (!string.IsNullOrEmpty(message.Error))
+            {
+                writer.WriteString(ErrorPropertyNameBytes, message.Error);
+            }
+            else if (message.HasResult)
+            {
+                var bytes = JsonSerializer.ToBytes(message.Result, message.Result?.GetType());
+                using var token = JsonDocument.Parse(bytes);
+                token.RootElement.WriteAsProperty(ResultPropertyNameBytes, ref writer);
+            }
+        }
+
+        private void WriteCancelInvocationMessage(CancelInvocationMessage message, ref Utf8JsonWriter writer)
+        {
+            WriteInvocationId(message, ref writer);
+        }
+
+        private void WriteStreamItemMessage(StreamItemMessage message, ref Utf8JsonWriter writer)
+        {
+            WriteInvocationId(message, ref writer);
+
+            var bytes = JsonSerializer.ToBytes(message.Item, message.Item?.GetType());
+            using var token = JsonDocument.Parse(bytes);
+            token.RootElement.WriteAsProperty(ItemPropertyNameBytes, ref writer);
+        }
+
+        private void WriteInvocationMessage(InvocationMessage message, ref Utf8JsonWriter writer)
+        {
+            WriteInvocationId(message, ref writer);
+            writer.WriteString(TargetPropertyNameBytes, message.Target);
+
+            WriteArguments(message.Arguments, ref writer);
+
+            WriteStreamIds(message.StreamIds, ref writer);
+        }
+
+        private void WriteStreamInvocationMessage(StreamInvocationMessage message, ref Utf8JsonWriter writer)
+        {
+            WriteInvocationId(message, ref writer);
+            writer.WriteString(TargetPropertyNameBytes, message.Target);
+
+            WriteArguments(message.Arguments, ref writer);
+
+            WriteStreamIds(message.StreamIds, ref writer);
+        }
+
+        private void WriteCloseMessage(CloseMessage message, ref Utf8JsonWriter writer)
+        {
+            if (message.Error != null)
+            {
+                writer.WriteString(ErrorPropertyNameBytes, message.Error);
+            }
+        }
+
+        private void WriteArguments(object[] arguments, ref Utf8JsonWriter writer)
+        {
+            writer.WriteStartArray(ArgumentsPropertyNameBytes);
+            foreach (var argument in arguments)
+            {
+                var type = argument?.GetType();
+                if (type == typeof(DateTime))
+                {
+                    writer.WriteStringValue((DateTime)argument);
+                }
+                else if (type == typeof(DateTimeOffset))
+                {
+                    writer.WriteStringValue((DateTimeOffset)argument);
+                }
+                else
+                {
+                    var bytes = JsonSerializer.ToBytes(argument, type);
+                    using var token = JsonDocument.Parse(bytes);
+                    token.RootElement.WriteAsValue(ref writer);
```

Comment:
We still need to be able to serialize directly into a writer, but I does Utf8JsonWriter have a concept similar Json.NET's WriteRawValue? We'd still allocate an array, but least then we wouldn't need to serialize, parse then serialize again.

Reply:
> does Utf8JsonWriter have a concept similar Json.NET's WriteRawValue

No.  "really raw" violates indentation settings, can violate the state (opening an array or object it doesn't close, closing the existing array or object), and can violate the character escaping rules for the writer.

So it's something that's hard to add... because it either isn't "raw" or leads to unexpected data.  If it's a token-by-token thing then there's the possibility that it writes stuff before finding out that the payload is invalid...  writing through JsonDocument solves the validity, escaping, and indentation problems; so it's the approach that we are encouraging for now.

Reply:
Serializing to bytes, then wrapping in a JsonDocument, then writing the JsonDocument is needlessly complex.

JsonSerializer should support serializing directly to a Utf8JsonWriter.

Reply:
> JsonSerializer should support serializing directly to a Utf8JsonWriter.

Oh, yeah, I agree.  I believe it's on the serializer TODO list (though if you can't find an issue then opening one would have virtue).  I was mainly commenting on WriteRaw for the general case.

Reply:
ok!

Reply:
Filed https://github.com/dotnet/corefx/issues/36714 as the closest thing I could find was the `Utf8JsonWriter` issue talking about `WriteRaw`

Reply:
Do you need equivalent functionality on the reader side? If so, an issue for that would be most welcome :)

I consider this under the umbrella of "interop between serializer and reader/writer" (similar to what we recently added - interop between jsondocument and reader/writer - i.e. https://github.com/dotnet/corefx/issues/36148 / https://github.com/dotnet/corefx/issues/36153).

Reply:
Yes we do, I'm currently looking at this and can hopefully show some perf numbers for the difference between using the JsonDocument approach we currently have vs. using the raw sequence.

Reply:
Filed for that case https://github.com/dotnet/corefx/issues/36717